File: web/src/api/catalog.ts
import { api } from "./client";
import {
  toProduct,
  toProductCard,
  type ApiProductCard,
  type ApiProductDetail,
} from "./adapters/productAdapter";
import type { Paginated, Product, ProductFilters, ProductSort } from "@/types/product";
import type { PincodeLookup } from "@/types/customer";
import type { Review } from "@/types/review";

/**
 * Catalogue API.
 *
 * Takes and returns furniture domain types. Every filter in `ProductFilters`
 * is now genuinely forwarded and genuinely applied — the backend supports the
 * facets the UI was architected for, so nothing is dropped on the way out and
 * no filter chip claims to be narrowing a grid it did not narrow.
 */

export interface Category {
  id: string;
  slug: string;
  name: string;
  imageUrl: string | null;
  productCount: number;
  children?: Category[];
}

export interface Room {
  id: string;
  slug: string;
  name: string;
  imageUrl: string | null;
}

export interface Collection {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  imageUrl: string | null;
  productCount: number;
  isFeatured: boolean;
}

export interface FilterOptions {
  attributes: {
    material?: AttributeOption[];
    finish?: AttributeOption[];
    colour?: AttributeOption[];
    style?: AttributeOption[];
    upholstery?: AttributeOption[];
  };
  brands: Array<{ id: string; slug: string; name: string }>;
  price: { min: number; max: number };
}

export interface AttributeOption {
  id: string;
  slug: string;
  name: string;
  hexCode: string | null;
}

export interface SearchSuggestions {
  query: string;
  products: Array<{
    id: string;
    slug: string;
    name: string;
    imageUrl: string | null;
    price: number;
  }>;
  categories: Array<{ id: string; name: string; slug: string }>;
  brands: Array<{ id: string; name: string }>;
}

interface ApiPage<T> {
  items: T[];
  total: number;
  limit: number;
  offset: number;
  has_more: boolean;
}

interface ApiCategoryNode {
  id: number;
  parent_id: number | null;
  name: string;
  slug: string;
  image_url: string | null;
  product_count: number;
  children: ApiCategoryNode[];
}

const DEFAULT_LIMIT = 24;

/** The sorts the API accepts. `discount` is not one of them. */
const SORTS: Record<ProductSort, string | null> = {
  newest: "newest",
  price_asc: "price_asc",
  price_desc: "price_desc",
  rating: "rating",
  // No discount sort exists. Clamped to the default rather than sent and
  // silently ignored, which would leave the chip lit over an unsorted grid.
  discount: null,
};

function page<T>(raw: ApiPage<unknown>, map: (item: never) => T): Paginated<T> {
  return {
    items: (raw.items as never[]).map(map),
    total: raw.total,
    limit: raw.limit,
    offset: raw.offset,
    hasMore: raw.has_more,
  };
}

function toCategory(node: ApiCategoryNode): Category {
  return {
    id: String(node.id),
    slug: node.slug,
    name: node.name,
    imageUrl: node.image_url,
    productCount: node.product_count,
    children: node.children?.length ? node.children.map(toCategory) : undefined,
  };
}

export const catalogApi = {
  /**
   * Product listing — one endpoint behind the shop, a category page, a room
   * page and search.
   *
   * `categoryId` in `ProductFilters` carries a *slug*, because that is what
   * the URL segment holds and what the API filters on. The field keeps its
   * name so no component or URL-state helper has to change.
   */
  async listProducts(
    filters: ProductFilters = {},
    limit = DEFAULT_LIMIT,
    offset = 0,
    signal?: AbortSignal,
  ): Promise<Paginated<Product>> {
    const params: Record<string, string | number | boolean | undefined> = {
      q: filters.query,
      category: filters.categoryId,
      min_price: filters.minPrice,
      max_price: filters.maxPrice,
      in_stock: filters.inStockOnly ? true : undefined,
      sort: filters.sort ? (SORTS[filters.sort] ?? undefined) : undefined,
      limit,
      offset,
    };

    const search = new URLSearchParams();
    for (const [key, value] of Object.entries(params)) {
      if (value !== undefined && value !== null && value !== "") {
        search.set(key, String(value));
      }
    }

    // Repeatable parameters: the API reads several values per facet and ORs
    // them, so `material=oak&material=walnut` means "oak or walnut". A
    // comma-joined single value would be read as one attribute slug that
    // matches nothing.
    const repeated: Array<[string, string[] | undefined]> = [
      ["material", filters.material],
      ["colour", filters.colour],
      ["style", filters.style],
      ["room", filters.room],
    ];
    for (const [key, values] of repeated) {
      for (const value of values ?? []) search.append(key, value);
    }
    for (const seats of filters.seatingCapacity ?? []) {
      search.append("seats", String(seats));
    }

    const raw = await api.get<ApiPage<ApiProductCard>>(
      `/catalog/products?${search.toString()}`,
      { auth: false, signal },
    );
    return page(raw, toProductCard as (item: never) => Product);
  },

  async getProduct(slug: string): Promise<Product> {
    const raw = await api.get<ApiProductDetail>(
      `/catalog/products/${encodeURIComponent(slug)}`,
    );
    return toProduct(raw);
  },

  /** "More like this" — same category, excluding the piece itself. */
  async getSimilarProducts(slug: string, limit = 12): Promise<Product[]> {
    const raw = await api.get<ApiProductCard[]>(
      `/catalog/products/${encodeURIComponent(slug)}/related`,
      { auth: false, params: { limit } },
    );
    return (raw ?? []).map(toProductCard);
  },

  /**
   * A view is recorded by the product endpoint itself when a customer is
   * signed in, so there is nothing separate to fire. Kept as a no-op rather
   * than removed, so the PDP's effect does not need a conditional.
   */
  async trackView(): Promise<void> {
    /* recorded server-side by GET /catalog/products/{slug} */
  },

  async getRecentlyViewed(limit = 8): Promise<Product[]> {
    const raw = await api.get<ApiProductCard[]>("/account/recently-viewed", {
      params: { limit },
    });
    return (raw ?? []).map(toProductCard);
  },

  /** The full category tree, with counts. */
  async getCategories(): Promise<Category[]> {
    const raw = await api.get<ApiCategoryNode[]>("/catalog/categories", {
      auth: false,
      next: { revalidate: 300 },
    });
    return (raw ?? []).map(toCategory);
  },

  async getRooms(): Promise<Room[]> {
    const raw = await api.get<
      Array<{ id: number; slug: string; name: string; image_url: string | null }>
    >("/catalog/rooms", { auth: false, next: { revalidate: 300 } });
    return (raw ?? []).map((r) => ({
      id: String(r.id),
      slug: r.slug,
      name: r.name,
      imageUrl: r.image_url,
    }));
  },

  async getCollections(featuredOnly = false): Promise<Collection[]> {
    const raw = await api.get<
      Array<{
        id: number;
        slug: string;
        name: string;
        description: string | null;
        image_url: string | null;
        product_count: number;
        is_featured: boolean;
      }>
    >("/catalog/collections", {
      auth: false,
      params: { featured_only: featuredOnly || undefined },
      next: { revalidate: 300 },
    });
    return (raw ?? []).map((c) => ({
      id: String(c.id),
      slug: c.slug,
      name: c.name,
      description: c.description,
      imageUrl: c.image_url,
      productCount: c.product_count,
      isFeatured: c.is_featured,
    }));
  },

  /**
   * A collection's products.
   *
   * This is the product list narrowed by `collection`, not a separate
   * endpoint, so a collection page gets real totals, real paging and the same
   * filters every other listing has — none of which the old dedicated
   * endpoint could offer.
   */
  async getCollection(
    slug: string,
    limit = DEFAULT_LIMIT,
    offset = 0,
    query?: string,
  ): Promise<{ name: string; products: Paginated<Product> }> {
    const [collections, raw] = await Promise.all([
      this.getCollections(),
      api.get<ApiPage<ApiProductCard>>("/catalog/products", {
        auth: false,
        params: { collection: slug, q: query, limit, offset },
      }),
    ]);

    const match = collections.find((c) => c.slug === slug);
    // A retired campaign URL should 404, not render an empty grid under a
    // heading that implies the collection still exists.
    if (!match) {
      const error = new Error("No such collection");
      (error as Error & { status?: number }).status = 404;
      throw error;
    }

    return {
      name: match.name,
      products: page(raw, toProductCard as (item: never) => Product),
    };
  },

  /** Everything a filter panel needs, in one call. */
  async getFilterOptions(): Promise<FilterOptions> {
    const raw = await api.get<{
      attributes: Record<
        string,
        Array<{ id: number; slug: string; name: string; hex_code: string | null }>
      >;
      brands: Array<{ id: number; slug: string; name: string }>;
      price: { min: number; max: number };
    }>("/catalog/filters", { auth: false, next: { revalidate: 300 } });

    const attributes: FilterOptions["attributes"] = {};
    for (const [kind, values] of Object.entries(raw?.attributes ?? {})) {
      attributes[kind as keyof FilterOptions["attributes"]] = values.map((v) => ({
        id: String(v.id),
        slug: v.slug,
        name: v.name,
        hexCode: v.hex_code,
      }));
    }

    return {
      attributes,
      brands: (raw?.brands ?? []).map((b) => ({
        id: String(b.id),
        slug: b.slug,
        name: b.name,
      })),
      price: raw?.price ?? { min: 0, max: 0 },
    };
  },

  /**
   * Search suggestions.
   *
   * Built from the product list rather than a dedicated endpoint, because the
   * list already does substring matching over name, tagline and description.
   * A separate suggestions endpoint would be a second definition of "what
   * matches", and the two would drift.
   */
  async getSearchSuggestions(
    query: string,
    limit = 6,
    signal?: AbortSignal,
  ): Promise<SearchSuggestions> {
    const raw = await api.get<ApiPage<ApiProductCard>>("/catalog/products", {
      auth: false,
      params: { q: query, limit },
      signal,
    });

    const products = (raw?.items ?? []).map((item) => ({
      id: String(item.id),
      slug: item.slug,
      name: item.name,
      imageUrl: item.primary_image?.url ?? null,
      price: Number(item.price_from) || 0,
    }));

    // Categories and brands are derived from what actually matched, so every
    // suggestion leads somewhere with results in it.
    const categories = new Map<string, { id: string; name: string; slug: string }>();
    const brands = new Map<string, { id: string; name: string }>();
    for (const item of raw?.items ?? []) {
      if (item.category) {
        categories.set(item.category.slug, {
          id: String(item.category.id),
          name: item.category.name,
          slug: item.category.slug,
        });
      }
      if (item.brand) {
        brands.set(String(item.brand.id), {
          id: String(item.brand.id),
          name: item.brand.name,
        });
      }
    }

    return {
      query,
      products,
      categories: [...categories.values()],
      brands: [...brands.values()],
    };
  },

  /* ----------------------------------------------------------------- Wishlist */

  async getWishlist(): Promise<Product[]> {
    const raw = await api.get<ApiProductCard[]>("/account/wishlist");
    return (raw ?? []).map(toProductCard);
  },

  async addToWishlist(productId: string): Promise<void> {
    await api.post(`/account/wishlist/${encodeURIComponent(productId)}`);
  },

  async removeFromWishlist(productId: string): Promise<void> {
    await api.del(`/account/wishlist/${encodeURIComponent(productId)}`);
  },

  /* ------------------------------------------------------------------ Reviews */

  async listReviews(productId: string, limit = 20, offset = 0): Promise<Review[]> {
    const raw = await api.get<
      ApiPage<{
        id: number;
        rating: number;
        title: string | null;
        body: string | null;
        author_name: string;
        is_verified_purchase: boolean;
        created_at: string;
      }>
    >(`/products/${encodeURIComponent(productId)}/reviews`, {
      auth: false,
      params: { limit, offset },
    });

    return (raw?.items ?? []).map((r) => ({
      id: String(r.id),
      rating: r.rating,
      // Title and body are separate fields on the wire; the UI shows one
      // block, so they are joined here rather than in a component.
      comment: [r.title, r.body].filter(Boolean).join(" — ") || null,
      authorName: r.author_name,
      createdAt: r.created_at,
      isVerifiedPurchase: r.is_verified_purchase,
    }));
  },

  /**
   * The star distribution, as a server-side aggregate.
   *
   * Computed over every approved review, not just the page that happens to be
   * loaded — a client-side histogram silently describes only the first ten.
   */
  async getReviewSummary(productId: string): Promise<{
    average: number | null;
    count: number;
    distribution: Record<1 | 2 | 3 | 4 | 5, number>;
  }> {
    const raw = await api.get<{
      average: number | null;
      count: number;
      distribution: Record<string, number>;
    }>(`/products/${encodeURIComponent(productId)}/reviews/summary`, { auth: false });

    return {
      average: raw?.average ?? null,
      count: raw?.count ?? 0,
      distribution: {
        1: raw?.distribution?.["1"] ?? 0,
        2: raw?.distribution?.["2"] ?? 0,
        3: raw?.distribution?.["3"] ?? 0,
        4: raw?.distribution?.["4"] ?? 0,
        5: raw?.distribution?.["5"] ?? 0,
      },
    };
  },

  async submitReview(
    productId: string,
    input: { rating: number; comment?: string; authorName?: string },
  ): Promise<void> {
    await api.post(`/account/reviews/${encodeURIComponent(productId)}`, {
      rating: input.rating,
      body: input.comment,
      author_name: input.authorName,
    });
  },

  /* ------------------------------------------------------------------ Delivery */

  /**
   * Serviceability for a PIN code.
   *
   * Answers what a furniture buyer is actually asking — will you deliver,
   * when, and for how much — from the same shipping zones that will price the
   * order at checkout, so the estimate cannot disagree with the bill.
   *
   * City and state stay null: no PIN database backs this, and guessing a city
   * would put a wrong one under a customer's address. Resolves to null on any
   * failure so a lookup miss can never block a checkout.
   */
  async lookupPincode(pincode: string): Promise<PincodeLookup | null> {
    try {
      const raw = await api.get<{
        pincode: string;
        serviceable: boolean;
        zone: string | null;
        shipping_fee?: number;
        free_above?: number | null;
        estimated_days_min?: number | null;
        estimated_days_max?: number | null;
      }>(`/catalog/serviceability/${encodeURIComponent(pincode)}`, { auth: false });

      return {
        pincode,
        city: null,
        state: null,
        serviceable: raw.serviceable,
        zone: raw.zone,
        shippingFee: raw.shipping_fee ?? null,
        freeAbove: raw.free_above ?? null,
        estimatedDaysMin: raw.estimated_days_min ?? null,
        estimatedDaysMax: raw.estimated_days_max ?? null,
      };
    } catch {
      return null;
    }
  },
};

File: apis/app/storefront/routes/catalog.py
"""The public catalogue: browse, filter, search, product detail."""
from decimal import Decimal

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import and_, exists, func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.config import settings
from app.core.database import get_db
from app.core.rbac import get_optional_customer
from app.models.catalog import (
    Attribute,
    Brand,
    Category,
    Collection,
    CollectionProduct,
    Product,
    ProductAttribute,
    ProductRoom,
    ProductVariant,
    RecentlyViewed,
    Review,
    Room,
)
from app.models.customer import Customer
from app.schemas.catalog import (
    AttributeOut,
    BrandOut,
    CategoryTree,
    CollectionOut,
    ProductCard,
    ProductDetail,
    RoomOut,
)
from app.schemas.common import Page
from app.services import catalog as catalog_service
from app.services import catalog_supabase

router = APIRouter(prefix="/catalog", tags=["Shop · Catalogue"])

_LOADED = (
    selectinload(Product.variants),
    selectinload(Product.images),
    selectinload(Product.category),
    selectinload(Product.brand),
)

# Only active products are ever visible. Drafts and archived rows are staff
# concepts; a customer meeting one is a leak, not a feature.
_VISIBLE = Product.status == "active"


@router.get("/products", response_model=Page[ProductCard])
async def list_products(
    q: str | None = Query(None, min_length=2, description="Free-text search"),
    category: str | None = Query(None, description="Category slug"),
    room: str | None = Query(None, description="Room slug"),
    collection: str | None = Query(None, description="Collection slug"),
    brand: str | None = None,
    material: list[str] | None = Query(None),
    finish: list[str] | None = Query(None),
    colour: list[str] | None = Query(None),
    style: list[str] | None = Query(None),
    min_price: Decimal | None = Query(None, ge=0),
    max_price: Decimal | None = Query(None, ge=0),
    max_width_mm: int | None = Query(None, ge=0, description="Will it fit through the door"),
    seats: int | None = Query(None, ge=1, le=20),
    in_stock: bool | None = None,
    sort: str = Query("featured", description="featured | price_asc | price_desc | newest | rating"),
    limit: int = Query(24, ge=1, le=48),
    offset: int = Query(0, ge=0),
    db: AsyncSession = Depends(get_db),
):
    """One endpoint for the shop, a category page, a room page and search.

    Every filter listed here is genuinely applied. A filter the backend
    cannot honour is not accepted and silently ignored - a chip that claims
    to be narrowing an unfiltered grid is a lie the customer can see.
    """
    query = select(Product).options(*_LOADED).where(_VISIBLE)
    count_query = select(func.count(func.distinct(Product.id))).select_from(Product).where(_VISIBLE)
    conditions = []

    if q:
        term = f"%{q.strip()}%"
        conditions.append(or_(Product.name.ilike(term), Product.tagline.ilike(term), Product.description.ilike(term)))

    if category:
        # Includes descendants: browsing "Seating" must show sofas, which sit
        # on the leaf, not on the parent.
        root = (await db.execute(select(Category).where(Category.slug == category))).scalars().first()
        if root is None:
            raise HTTPException(status.HTTP_404_NOT_FOUND, "No such category.")
        ids = {root.id}
        frontier = [root.id]
        while frontier:
            children = (
                await db.execute(select(Category.id).where(Category.parent_id.in_(frontier)))
            ).scalars().all()
            frontier = [c for c in children if c not in ids]
            ids.update(frontier)
        conditions.append(Product.category_id.in_(ids))

    if room:
        conditions.append(
            exists().where(
                and_(
                    ProductRoom.product_id == Product.id,
                    ProductRoom.room_id == select(Room.id).where(Room.slug == room).scalar_subquery(),
                )
            )
        )

    if collection:
        conditions.append(
            exists().where(
                and_(
                    CollectionProduct.product_id == Product.id,
                    CollectionProduct.collection_id
                    == select(Collection.id).where(Collection.slug == collection).scalar_subquery(),
                )
            )
        )

    if brand:
        conditions.append(
            Product.brand_id == select(Brand.id).where(Brand.slug == brand).scalar_subquery()
        )

    # Attribute facets. Within one facet the slugs are OR-ed (oak *or*
    # walnut); across facets they are AND-ed (oak *and* fabric) - which is
    # what a shopper means by ticking two boxes in two different lists.
    for kind, values in (("material", material), ("finish", finish), ("colour", colour), ("style", style)):
        if values:
            conditions.append(
                exists().where(
                    and_(
                        ProductAttribute.product_id == Product.id,
                        ProductAttribute.attribute_id.in_(
                            select(Attribute.id).where(Attribute.kind == kind, Attribute.slug.in_(values))
                        ),
                    )
                )
            )

    variant_conditions = [ProductVariant.is_active.is_(True)]
    if min_price is not None:
        variant_conditions.append(ProductVariant.price >= min_price)
    if max_price is not None:
        variant_conditions.append(ProductVariant.price <= max_price)
    if max_width_mm is not None:
        variant_conditions.append(ProductVariant.width_mm <= max_width_mm)
    if in_stock:
        variant_conditions.append(
            or_(ProductVariant.stock_quantity > 0, ProductVariant.backorder_allowed.is_(True))
        )
    if len(variant_conditions) > 1:
        # The condition is "some variant matches all of these", not "some
        # variant is under the max price and some other one is in stock".
        conditions.append(
            exists().where(and_(ProductVariant.product_id == Product.id, *variant_conditions))
        )

    if seats is not None:
        conditions.append(Product.seating_capacity >= seats)

    if conditions:
        query = query.where(*conditions)
        count_query = count_query.where(*conditions)

    total = await db.scalar(count_query) or 0

    if sort in ("price_asc", "price_desc"):
        # Order by the cheapest active variant, which is the price the card
        # actually shows. Joining and ordering by ProductVariant.price
        # directly would order by whichever row the join happened to pick.
        cheapest = (
            select(func.min(ProductVariant.price))
            .where(ProductVariant.product_id == Product.id, ProductVariant.is_active.is_(True))
            .scalar_subquery()
        )
        query = query.order_by(cheapest.asc() if sort == "price_asc" else cheapest.desc())
    elif sort == "newest":
        query = query.order_by(Product.created_at.desc())
    elif sort == "rating":
        average = (
            select(func.avg(Review.rating))
            .where(Review.product_id == Product.id, Review.status == "approved")
            .scalar_subquery()
        )
        query = query.order_by(average.desc().nullslast())
    else:
        query = query.order_by(Product.created_at.desc())

    rows = (await db.execute(query.limit(limit).offset(offset))).scalars().unique().all()
    ratings = await catalog_service.rating_map(db, [p.id for p in rows])

    return Page[ProductCard](
        items=[catalog_service.to_card(p, ratings) for p in rows],
        total=total, limit=limit, offset=offset,
    )


@router.get("/products/{slug}", response_model=ProductDetail)
async def get_product(
    slug: str,
    db: AsyncSession = Depends(get_db),
    customer: Customer | None = Depends(get_optional_customer),
):
    product = (
        await db.execute(
            select(Product)
            .options(*_LOADED, selectinload(Product.rooms), selectinload(Product.attributes))
            .where(Product.slug == slug, _VISIBLE)
        )
    ).scalars().first()
    if product is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "We could not find that piece.")

    if customer:
        existing = (
            await db.execute(
                select(RecentlyViewed).where(
                    RecentlyViewed.customer_id == customer.id, RecentlyViewed.product_id == product.id
                )
            )
        ).scalars().first()
        if existing:
            # Touch rather than insert, so "recently viewed" is ordered by
            # the most recent look rather than the first one.
            existing.updated_at = func.now()
        else:
            db.add(RecentlyViewed(customer_id=customer.id, product_id=product.id))
        await db.commit()

    from app.models.ar import ProductArAsset
    from app.services import ar as ar_service

    ratings = await catalog_service.rating_map(db, [product.id])
    asset = (
        await db.execute(
            select(ProductArAsset).where(ProductArAsset.product_id == product.id)
        )
    ).scalars().first()

    # to_public returns None for anything that is not a published, validated
    # model, so an unfinished or failed asset can never reach the storefront.
    return catalog_service.to_detail(
        product, ratings.get(product.id), ar=ar_service.to_public(asset)
    )


@router.get("/products/{slug}/related", response_model=list[ProductCard])
async def related_products(
    slug: str,
    limit: int = Query(8, ge=1, le=16),
    db: AsyncSession = Depends(get_db),
):
    """Same category, excluding the piece itself.

    Not a recommendation engine, and not presented as one - it is a "more
    like this" rail, which is what the data actually supports.
    """
    product = (await db.execute(select(Product).where(Product.slug == slug, _VISIBLE))).scalars().first()
    if product is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "We could not find that piece.")

    rows = (
        await db.execute(
            select(Product).options(*_LOADED)
            .where(_VISIBLE, Product.id != product.id, Product.category_id == product.category_id)
            .order_by(func.random()).limit(limit)
        )
    ).scalars().unique().all()
    ratings = await catalog_service.rating_map(db, [p.id for p in rows])
    return [catalog_service.to_card(p, ratings) for p in rows]


@router.get("/categories", response_model=list[CategoryTree])
async def categories(db: AsyncSession = Depends(get_db)):
    # The database may not be reachable by the wire protocol at all - see
    # docs/CPANEL-SUPABASE-HTTP.md. `db` is still injected and simply unused
    # on that path; removing the dependency would change the signature for
    # both backends to save one unopened connection.
    if settings.DATA_BACKEND == "supabase":
        return await catalog_supabase.category_tree()

    rows = (
        await db.execute(
            select(Category).where(Category.is_active.is_(True)).order_by(Category.position, Category.name)
        )
    ).scalars().all()
    counts = dict(
        (await db.execute(
            select(Product.category_id, func.count(Product.id))
            .where(_VISIBLE).group_by(Product.category_id)
        )).all()
    )
    nodes = {
        r.id: CategoryTree.model_validate({**r.__dict__, "children": [], "product_count": counts.get(r.id, 0)})
        for r in rows
    }
    roots = []
    for row in rows:
        parent = nodes.get(row.parent_id) if row.parent_id else None
        (parent.children if parent else roots).append(nodes[row.id])
    return roots


@router.get("/rooms", response_model=list[RoomOut])
async def rooms(db: AsyncSession = Depends(get_db)):
    if settings.DATA_BACKEND == "supabase":
        return await catalog_supabase.rooms()

    rows = (
        await db.execute(select(Room).where(Room.is_active.is_(True)).order_by(Room.position, Room.name))
    ).scalars().all()
    return [RoomOut.model_validate(r) for r in rows]


@router.get("/collections", response_model=list[CollectionOut])
async def collections(
    featured_only: bool = False,
    db: AsyncSession = Depends(get_db),
):
    query = select(Collection).where(Collection.is_active.is_(True))
    if featured_only:
        query = query.where(Collection.is_featured.is_(True))
    rows = (await db.execute(query.order_by(Collection.position, Collection.name))).scalars().all()
    counts = dict(
        (await db.execute(
            select(CollectionProduct.collection_id, func.count(CollectionProduct.product_id))
            .group_by(CollectionProduct.collection_id)
        )).all()
    )
    return [
        CollectionOut.model_validate({**r.__dict__, "product_count": counts.get(r.id, 0)})
        for r in rows
    ]


@router.get("/filters")
async def filter_options(db: AsyncSession = Depends(get_db)):
    """Everything a filter panel needs, in one call.

    Served rather than hardcoded in the frontend so a finish added in the
    dashboard appears as a filter without a frontend release. Only
    attributes that are actually on a visible product are returned - a
    filter that can only ever return nothing is worse than no filter.
    """
    used = (
        select(ProductAttribute.attribute_id)
        .join(Product, Product.id == ProductAttribute.product_id)
        .where(_VISIBLE)
    )
    attributes = (
        await db.execute(
            select(Attribute)
            .where(Attribute.is_active.is_(True), Attribute.id.in_(used))
            .order_by(Attribute.kind, Attribute.position, Attribute.name)
        )
    ).scalars().all()

    price_range = (
        await db.execute(
            select(func.min(ProductVariant.price), func.max(ProductVariant.price))
            .join(Product, Product.id == ProductVariant.product_id)
            .where(_VISIBLE, ProductVariant.is_active.is_(True))
        )
    ).first()

    brands = (
        await db.execute(
            select(Brand).where(
                Brand.is_active.is_(True),
                Brand.id.in_(select(Product.brand_id).where(_VISIBLE, Product.brand_id.isnot(None))),
            ).order_by(Brand.name)
        )
    ).scalars().all()

    grouped: dict[str, list] = {}
    for attribute in attributes:
        grouped.setdefault(attribute.kind, []).append(AttributeOut.model_validate(attribute).model_dump())

    return {
        "attributes": grouped,
        "brands": [BrandOut.model_validate(b).model_dump() for b in brands],
        "price": {
            "min": float(price_range[0]) if price_range and price_range[0] is not None else 0,
            "max": float(price_range[1]) if price_range and price_range[1] is not None else 0,
        },
    }


@router.get("/serviceability/{pincode}")
async def serviceability(pincode: str, db: AsyncSession = Depends(get_db)):
    """Can we deliver here, how long will it take, and what does it cost.

    Answered from the shipping zones staff actually maintain, so the estimate
    on a product page is the same rule that will price the order at checkout.
    A city/state lookup would need a PIN-code database this project does not
    have, and inventing one would put a wrong city under a customer's address.
    """
    if not (pincode.isdigit() and len(pincode) == 6):
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Enter a six-digit PIN code.")

    from app.models.commerce import ShippingRate

    rates = (
        await db.execute(
            select(ShippingRate).where(ShippingRate.is_active.is_(True)).order_by(ShippingRate.position)
        )
    ).scalars().all()

    chosen = None
    best = -1
    fallback = None
    for rate in rates:
        prefixes = [p.strip() for p in (rate.postcode_prefixes or "").split(",") if p.strip()]
        if not prefixes:
            fallback = fallback or rate
            continue
        for prefix in prefixes:
            if pincode.startswith(prefix) and len(prefix) > best:
                chosen, best = rate, len(prefix)

    rate = chosen or fallback
    if rate is None:
        # No zone at all is a configuration gap, not a refusal to deliver.
        # Saying "we do not deliver there" would turn a missing row into a
        # lost sale.
        return {"pincode": pincode, "serviceable": True, "zone": None}

    return {
        "pincode": pincode,
        "serviceable": True,
        "zone": rate.name,
        "shipping_fee": float(rate.rate),
        "free_above": float(rate.free_above) if rate.free_above is not None else None,
        "estimated_days_min": rate.estimated_days_min,
        "estimated_days_max": rate.estimated_days_max,
    }
